src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -173,6 +182,9 @@ public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws
                     }
                     remainingMaxTimeToWait =
                             poolNotLocked.awaitNanos(remainingMaxTimeToWait);
+                } else if (this.poisonPill != null){
+                    // Return the pill immediately if there is one
+                    item = poisonPill;
```

Comment:
Same question here about whether this could be done before the pool lock await above.